cpp/src/plasma/external_store.h

```diff
@@ -0,0 +1,124 @@
+// Licensed to the Apache Software Foundation (ASF) under one
+// or more contributor license agreements.  See the NOTICE file
+// distributed with this work for additional information
+// regarding copyright ownership.  The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License.  You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+#ifndef EXTERNAL_STORE_H
+#define EXTERNAL_STORE_H
+
+#include <map>
+#include "plasma/client.h"
+
+namespace plasma {
+
+// ==== The external store ====
+//
+// This file contains declaration for all functions that need to be implemented
+// for an external storage service so that objects evicted from Plasma store
+// can be written to it.
+
+class ExternalStoreHandle {
+ public:
+  ExternalStoreHandle() = default;
+
+  virtual ~ExternalStoreHandle() = default;
+
+  /// This method will be called whenever an object in the Plasma store needs
+  /// to be evicted to the external store.
+  ///
+  /// \param num_objects The number of objects to put.
+  /// \param ids The IDs of the objects to put.
+  /// \param data The object data to put.
+  /// \return The return status.
+  virtual Status Put(size_t num_objects, const ObjectID *ids, const std::string *data) = 0;
```

Comment:
This part was a little confusing to me. Why isn't the signature taking a const std::vector<ObjectID>& and a const std::vector<std::string>&; then I saw it is used to put chunks and I guess we want to avoid copies of the data here.

I'd consider renaming it/making the signature clearer. I'd also consider using arrow::Buffer for the data, since that can avoid copies and can also be conveniently sliced into, see https://github.com/apache/arrow/blob/master/cpp/src/arrow/buffer.h.

Reply:
I see yeah that makes sense.

The one that would have been natural for me (and safer because there is no pointer arithmetic) is
```
Status Put(const std::vector<ObjectID>& object_ids, const std::vector<std::string>& data)
```
if it's possible to support that efficiently.

Reply:
Yes, that was my first choice too. It's possible to use this efficiently if we delegate the parallelism in `Put` to specific `ExternalStoreHandle` implementations -- would that make sense?

Reply:
I agree with your point in using arrow::Buffer, and I ended up with the following signature for Get() and Put():
```cpp
Status Put(const std::vector<ObjectID> &ids, const std::vector<std::shared_ptr<Buffer>> &data);

Status Get(const std::vector<ObjectID> &ids, std::vector<std::string> &data);
```
 
I've removed parallelism for Put right now, we can add that back in if Put becomes a bottleneck.